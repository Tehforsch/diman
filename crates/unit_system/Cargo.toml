[package]
name = "unit_system"
version = "0.1.0"
edition = "2021"
authors = [
    "Toni Peter <mail@tonipeter.de>",
]

[features]
glam = []
glam-vec2 = []
glam-dvec2 = []
glam-vec3 = []
glam-dvec3 = []

[dependencies]
syn = { version = "1.0", features = ["full"] }
quote = "1.0"
proc-macro2 = "1.0"

[lib]
proc-macro = true
