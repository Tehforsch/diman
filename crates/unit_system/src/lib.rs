#![feature(proc_macro_diagnostic)]

mod parse;
mod types;
mod expression;
// mod storage_types;
// mod codegen;

use syn::*;
use types::{Defs};

#[proc_macro]
pub fn unit_system(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let _defs = parse_macro_input!(item as Defs);
    // defs.code_gen().into()
    proc_macro::TokenStream::new()
}
