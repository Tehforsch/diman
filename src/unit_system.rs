#[macro_export]
macro_rules! unit_system {
    ($dimension: ident, $quantity: ident, $dimensionless_const: ident, $($const: ident, $quantity_name:ident, $($dimension_name: ident: $dimension_value: literal),*, {$($unit:ident, $factor:literal, $($unit_symbol:literal)?),*}),+) => {
        use paste::paste;
        pub const UNIT_NAMES: &[($dimension, &str, f64)] = &[
        $(
            $(
                $(
                    ($const, $unit_symbol, $factor),
                )*
            )*
        )*
        ];
        $(
            #[allow(clippy::needless_update)]
            pub const $const: $dimension = $dimension {
                $(
                    $dimension_name: $dimension_value,
                )*
                .. $dimensionless_const };
            pub type $quantity_name = $quantity<f64, $const>;
            paste!{
                pub type [<Vec $quantity_name>] = $quantity<glam::DVec2, $const>;
                pub type [<DVec2 $quantity_name>] = $quantity<glam::DVec2, $const>;
                pub type [<DVec3 $quantity_name>] = $quantity<glam::DVec3, $const>;
            }

            impl $quantity::<f64, $const> {
                $(
                    pub const fn $unit(v: f64) -> $quantity::<f64, $const> {
                        $quantity::<f64, $const>(v * $factor)
                    }

                )*
            }

            impl $quantity::<f32, $const> {
                $(
                    pub const fn $unit(v: f64) -> $quantity::<f32, $const> {
                        $quantity::<f32, $const>((v * ($factor as f64)) as f32)
                    }
                )*
            }

            impl $quantity<glam::DVec2, $const> {
                $(
                    pub fn $unit(x: f64, y: f64) -> $quantity::<glam::DVec2, $const> {
                        $quantity::<glam::DVec2, $const>(glam::DVec2::new(x, y) * $factor)
                    }
                )*
            }

            impl $quantity<glam::DVec3, $const> {
                $(
                    pub fn $unit(x: f64, y: f64, z: f64) -> $quantity::<glam::DVec3, $const> {
                        $quantity::<glam::DVec3, $const>(glam::DVec3::new(x, y, z) * $factor)
                    }
                )*
            }

            impl<S> $quantity<S, $const> where S: Div<f64, Output = S> {
                paste! {
                    $(
                        pub fn [<in_ $unit>](self) -> S {
                            self.0 / $factor
                        }
                    )*
                }
            }
        )*
    }
}
