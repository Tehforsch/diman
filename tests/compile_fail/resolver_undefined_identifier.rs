#![allow(incomplete_features)]
#![feature(generic_const_exprs, adt_const_params)]

use diman::unit_system;

unit_system!(
    quantity_type Quantity,
    dimension_type Dimension,
    dimension Length,
    unit (kilometers, "km") = 1000.0 * meters,
);

fn main() {}
