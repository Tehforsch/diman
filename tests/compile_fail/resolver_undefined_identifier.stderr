error: Undefined identifier "meters".
  --> tests/compile_fail/resolver_undefined_identifier.rs:11:40
   |
11 |     unit (kilometers, "km") = 1000.0 * meters,
   |                                        ^^^^^^
